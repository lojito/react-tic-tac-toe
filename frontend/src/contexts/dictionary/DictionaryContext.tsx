import { createContext } from "react";
import { Language } from "../../contexts/language/LanguageContext";
import data from "../../data/json/dictionary.json";

export type DictionaryEntry = Partial<
  Record<
    | "ABOUT"
    | "ABOUT_GAME_DEVELOPED_WITH"
    | "ABOUT_OTHER_HOME_PROJECTS"
    | "ABOUT_SOURCE_CODE"
    | "AUTH_FORM_EMAIL_LABEL"
    | "AUTH_FORM_EMAIL_PLACEHOLDER"
    | "AUTH_FORM_NAME_LABEL"
    | "AUTH_FORM_NAME_PLACEHOLDER"
    | "AUTH_FORM_PASSWORD_LABEL"
    | "AUTH_FORM_PASSWORD_PLACEHOLDER"
    | "AUTH_ERROR_SERVER_NOT_RESPONDING"
    | "AUTH_ERROR_VALIDATION_FAILED"
    | "AUTH_ERROR_UNKNOWN"
    | "BUTTON_BACK"
    | "BUTTON_DELETE_GAME"
    | "BUTTON_LOGOUT"
    | "BUTTON_PLAY"
    | "BUTTON_PLAY_AGAIN"
    | "BUTTON_SEE_GAME"
    | "BUTTON_SIGNIN"
    | "BUTTON_SIGNUP"
    | "CATEGORY"
    | "CATEGORY_ANIMALS"
    | "CATEGORY_FRUITS_AND_VEGETABLES"
    | "CATEGORY_GERMANY_LANDMARKS"
    | "CATEGORY_HAVANA_LANDMARKS"
    | "CATEGORY_MONTREAL_LANDMARKS"
    | "CATEGORY_PUPPIES"
    | "CATEGORY_SEINFELD"
    | "CATEGORY_SOCCER_PLAYERS"
    | "CATEGORY_SPAIN_LANDMARKS"
    | "CATEGORY_VANCOUVER_LANDMARKS"
    | "CONTACT"
    | "CONTACT_AUTHOR"
    | "CONTACT_EMAIL"
    | "CONTACT_GITHUB"
    | "CONTACT_LINKEDIN"
    | "FLAG_ENGLISH_HINT"
    | "FLAG_FRENCH_HINT"
    | "FLAG_SPANISH_HINT"
    | "GAME"
    | "GAME_OVER"
    | "GAME_STARTED_BY"
    | "GAME_STARTED_BY_COMPUTER"
    | "GAME_STARTED_BY_YOU"
    | "GAME_DELETING"
    | "GAME_DELETE_ERROR"
    | "GAME_SAVE_ERROR"
    | "GAME_SAVING"
    | "HISTORICAL"
    | "HISTORICAL_NO_DATA_YET"
    | "HISTORICAL_FETCHING_GAMES"
    | "HISTORICAL_FETCHING_GAME"
    | "HISTORICAL_FETCHING_GAMES_ERROR"
    | "HISTORICAL_FETCHING_GAME_ERROR"
    | "IMAGE_COMPUTER_LABEL"
    | "IMAGE_COMPUTER_TOOLTIP"
    | "IMAGE_REFRESH_TOOLTIP"
    | "IMAGE_USER_LABEL"
    | "IMAGE_USER_TOOLTIP"
    | "LOGIN"
    | "MESSAGE_DRAW"
    | "MESSAGE_LOST"
    | "MESSAGE_WON"
    | "PLAYING_LEVEL_EASY"
    | "PLAYING_LEVEL_LABEL"
    | "PLAYING_LEVEL_SMART"
    | "PLAYING_LEVEL_TOOLTIP"
    | "REPOSITORY"
    | "SIGNUP"
    | "WELCOME_BACK"
    | "START_GAME_LABEL"
    | "START_GAME_ANSWER_NO"
    | "START_GAME_TOOLTIP"
    | "START_GAME_ANSWER_YES",
    string
  >
>;

export type DictionaryContextValue = {
  dictionary: DictionaryEntry;
};

export const DictionaryContext = createContext<DictionaryContextValue>({
  dictionary: data[Language.English],
});
