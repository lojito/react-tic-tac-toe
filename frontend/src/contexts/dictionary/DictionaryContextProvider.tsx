import React, { FC, useContext } from "react";
import data from "../../data/json/dictionary.json";
import { LanguageContext } from "../language/LanguageContext";
import {
  DictionaryContext,
  DictionaryContextValue,
  DictionaryEntry,
} from "./DictionaryContext";

const DictionaryContextProvider: FC = ({ children }) => {
  const { language } = useContext(LanguageContext);

  const ctx: DictionaryContextValue = {
    dictionary: data[language] as DictionaryEntry,
  };

  return (
    <DictionaryContext.Provider value={ctx}>
      {children}
    </DictionaryContext.Provider>
  );
};

export default DictionaryContextProvider;
