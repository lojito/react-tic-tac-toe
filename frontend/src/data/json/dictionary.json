[
  {
    "ABOUT": "About",
    "ABOUT_GAME_DEVELOPED_WITH": "Game developed with:",
    "ABOUT_OTHER_HOME_PROJECTS": "Other projects:",
    "ABOUT_SOURCE_CODE": "Source Code:",
    "AUTH_FORM_EMAIL_LABEL": "E-mail *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "E-mail",
    "AUTH_FORM_NAME_LABEL": "Name *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Name",
    "AUTH_FORM_PASSWORD_LABEL": "Password *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Password",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "Server is not responding.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validation failed. Invalid e-mail or password.",
    "AUTH_ERROR_UNKNOWN": "There was an error while trying to login or signup in the user.",
    "BUTTON_BACK": "Back",
    "BUTTON_DELETE_GAME": "Delete Game",
    "BUTTON_LOGOUT": "Logout",
    "BUTTON_PLAY": "Play Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Play again",
    "BUTTON_SEE_GAME": "See Game",
    "BUTTON_SIGNIN": "Log in",
    "BUTTON_SIGNUP": "Sign up",
    "CATEGORY": "Category",
    "CATEGORY_ANIMALS": "Animals",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Fruits & Vegetables",
    "CATEGORY_GERMANY_LANDMARKS": "Germany landmarks",
    "CATEGORY_HAVANA_LANDMARKS": "Havana landmarks",
    "CATEGORY_MONTREAL_LANDMARKS": "Montreal landmarks",
    "CATEGORY_PUPPIES": "Puppies",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Soccer players",
    "CATEGORY_SPAIN_LANDMARKS": "Spain landmarks",
    "CATEGORY_VANCOUVER_LANDMARKS": "Vancouver landmarks",
    "CONTACT": "Contact",
    "CONTACT_AUTHOR": "Author:",
    "CONTACT_EMAIL": "E-mail:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "English",
    "FLAG_FRENCH_HINT": "French",
    "FLAG_SPANISH_HINT": "Spanish",
    "GAME": "Game",
    "GAME_OVER": "Game over",
    "GAME_STARTED_BY": "Started by",
    "GAME_STARTED_BY_COMPUTER": "Computer",
    "GAME_STARTED_BY_YOU": "You",
    "GAME_DELETING": "Removing the game data from the database...",
    "GAME_DELETE_ERROR": "An error ocurred while deleting the game from the database.",
    "GAME_SAVING": "Saving the game data...",
    "GAME_SAVE_ERROR": "An error ocurred while saving the game results into the database.",
    "HISTORICAL": "Historical",
    "HISTORICAL_NO_DATA_YET": "There is no historical data yet. Play some games and comeback.",
    "HISTORICAL_FETCHING_GAMES": "Fetching games...",
    "HISTORICAL_FETCHING_GAME": "Fetching game...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "An error ocurred while fetching the games from the database.",
    "HISTORICAL_FETCHING_GAME_ERROR": "An error ocurred while fetching the game from the database.",
    "IMAGE_COMPUTER_LABEL": "PC",
    "IMAGE_COMPUTER_TOOLTIP": "The computer will play with this image.",
    "IMAGE_REFRESH_TOOLTIP": "Click to play with a different image.",
    "IMAGE_USER_LABEL": "You",
    "IMAGE_USER_TOOLTIP": "You will play with this image.",
    "LOGIN": "Log in",
    "MESSAGE_DRAW": "It's a draw!",
    "MESSAGE_LOST": "You lost!",
    "MESSAGE_WON": "You won!",
    "PLAYING_LEVEL_EASY": "Easy",
    "PLAYING_LEVEL_LABEL": "Level",
    "PLAYING_LEVEL_SMART": "Smart",
    "PLAYING_LEVEL_TOOLTIP": "How much thinking the computer does before playing.",
    "REPOSITORY": "Repository",
    "SIGNUP": "Sign up",
    "WELCOME_BACK": "Welcome back",
    "START_GAME_LABEL": "Start game",
    "START_GAME_ANSWER_NO": "No",
    "START_GAME_TOOLTIP": "Who starts playing the game: You or the computer?",
    "START_GAME_ANSWER_YES": "Yes"
  },
  {
    "ABOUT": "Acerca de",
    "ABOUT_GAME_DEVELOPED_WITH": "Juego desarrollado con:",
    "ABOUT_OTHER_HOME_PROJECTS": "Otros projectos:",
    "ABOUT_SOURCE_CODE": "Código fuente:",
    "AUTH_FORM_EMAIL_LABEL": "Correo electrónico *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "Correo electrónico",
    "AUTH_FORM_NAME_LABEL": "Nombre *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Nombre",
    "AUTH_FORM_PASSWORD_LABEL": "Contraseña *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Contraseña",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "El servidor no está respondiendo.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validación fallida. Correo electrónico o contraseña no válida.",
    "AUTH_ERROR_UNKNOWN": "Se produjo un error al intentar iniciar sesión o registrar al usuario.",
    "BUTTON_BACK": "Regresar",
    "BUTTON_DELETE_GAME": "Eliminar la partida",
    "BUTTON_LOGOUT": "Cerrar sesión",
    "BUTTON_PLAY": "Juega al Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Jugar de nuevo",
    "BUTTON_SEE_GAME": "Ver la partida",
    "BUTTON_SIGNIN": "Iniciar sesión",
    "BUTTON_SIGNUP": "Inscribirse",
    "CATEGORY": "Categoría",
    "CATEGORY_ANIMALS": "Animales",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Frutas y vegetales",
    "CATEGORY_GERMANY_LANDMARKS": "Monumentos de Alemania",
    "CATEGORY_HAVANA_LANDMARKS": "Monumentos de La Habana",
    "CATEGORY_MONTREAL_LANDMARKS": "Monumentos de Montreal",
    "CATEGORY_PUPPIES": "Cachorros",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Jugadores de fútbol",
    "CATEGORY_SPAIN_LANDMARKS": "Monumentos de España",
    "CATEGORY_VANCOUVER_LANDMARKS": "Monumentos de Vancouver",
    "CONTACT": "Contacto",
    "CONTACT_AUTHOR": "Autor:",
    "CONTACT_EMAIL": "Correo electrónico:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "Inglés",
    "FLAG_FRENCH_HINT": "Francés",
    "FLAG_SPANISH_HINT": "Español",
    "GAME": "Juego",
    "GAME_OVER": "Partida acabada",
    "GAME_STARTED_BY": "Comenzada por",
    "GAME_STARTED_BY_COMPUTER": "Ordenador",
    "GAME_STARTED_BY_YOU": "Ti",
    "GAME_DELETING": "Eliminando el juego de la base de datos...",
    "GAME_DELETE_ERROR": "Ocurrió un error al eliminar el juego de la base de datos.",
    "GAME_SAVING": "Guardando los datos del juego...",
    "GAME_SAVE_ERROR": "Ocurrió un error al salvar los resultados del juego de la base de datos.",
    "HISTORICAL": "Histórico",
    "HISTORICAL_NO_DATA_YET": "No hay historial aún. Juega algunas partidas y regresa más tarde.",
    "HISTORICAL_FETCHING_GAMES": "Obteniendo juegos...",
    "HISTORICAL_FETCHING_GAME": "Obteniendo juego...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "Ocurrió un error al recuperar los juegos de la base de datos.",
    "HISTORICAL_FETCHING_GAME_ERROR": "Ocurrió un error al recuperar el juego de la base de datos.",
    "IMAGE_COMPUTER_LABEL": "PC",
    "IMAGE_COMPUTER_TOOLTIP": "El ordenador jugará con esta imagen.",
    "IMAGE_REFRESH_TOOLTIP": "Haz click para jugar con una imagen diferente.",
    "IMAGE_USER_LABEL": "Tú",
    "IMAGE_USER_TOOLTIP": "Tú jugaras con esta imagen.",
    "LOGIN": "Iniciar sesión",
    "MESSAGE_DRAW": "Empate!",
    "MESSAGE_LOST": "Perdiste!",
    "MESSAGE_WON": "Ganaste!",
    "PLAYING_LEVEL_EASY": "Fácil",
    "PLAYING_LEVEL_LABEL": "Nivel de juego",
    "PLAYING_LEVEL_SMART": "Difícil",
    "PLAYING_LEVEL_TOOLTIP": "Cuánto tiempo el ordenador piensa antes de jugar.",
    "REPOSITORY": "Banco",
    "SIGNUP": "Regístrate",
    "WELCOME_BACK": "Bienvenido de nuevo",
    "START_GAME_LABEL": "Tú empiezas",
    "START_GAME_ANSWER_NO": "No",
    "START_GAME_TOOLTIP": "Quién empieza la partida: Tú o el ordenador?",
    "START_GAME_ANSWER_YES": "Sí"
  },
  {
    "ABOUT": "À propos de",
    "ABOUT_GAME_DEVELOPED_WITH": "Jeu développé avec:",
    "ABOUT_OTHER_HOME_PROJECTS": "D'autres projects:",
    "ABOUT_SOURCE_CODE": "Code source:",
    "AUTH_FORM_EMAIL_LABEL": "Courrier électronique *",
    "AUTH_FORM_EMAIL_PLACEHOLDER": "Courrier électronique",
    "AUTH_FORM_NAME_LABEL": "Nom *",
    "AUTH_FORM_NAME_PLACEHOLDER": "Nom",
    "AUTH_FORM_PASSWORD_LABEL": "Mot de passe *",
    "AUTH_FORM_PASSWORD_PLACEHOLDER": "Mot de passe",
    "AUTH_ERROR_SERVER_NOT_RESPONDING": "Le serveur ne répond pas.",
    "AUTH_ERROR_VALIDATION_FAILED": "Validation échouée. Email ou mot de passe invalide.",
    "AUTH_ERROR_UNKNOWN": "Une erreur s'est produite lors de la tentative de connexion ou d'inscription de l'utilisateur.",
    "BUTTON_BACK": "Retour",
    "BUTTON_DELETE_GAME": "Retirer la partie",
    "BUTTON_LOGOUT": "Se déconnecter",
    "BUTTON_PLAY": "Joue au Tic-Tac-Toe",
    "BUTTON_PLAY_AGAIN": "Jouer à nouveau",
    "BUTTON_SEE_GAME": "Voir la partie",
    "BUTTON_SIGNIN": "Se connecter",
    "BUTTON_SIGNUP": "S'inscrire",
    "CATEGORY": "Catégorie",
    "CATEGORY_ANIMALS": "Des animaux",
    "CATEGORY_FRUITS_AND_VEGETABLES": "Fruits et légumes",
    "CATEGORY_GERMANY_LANDMARKS": "Monuments L'Allemagne",
    "CATEGORY_HAVANA_LANDMARKS": "Monuments de La Havane",
    "CATEGORY_MONTREAL_LANDMARKS": "Monuments de Montréal",
    "CATEGORY_PUPPIES": "Des chiots",
    "CATEGORY_SEINFELD": "Seinfeld",
    "CATEGORY_SOCCER_PLAYERS": "Joueurs du soccer",
    "CATEGORY_SPAIN_LANDMARKS": "Monuments L'Espagne",
    "CATEGORY_VANCOUVER_LANDMARKS": "Monuments de Vancouver",
    "CONTACT": "Contact",
    "CONTACT_AUTHOR": "Auteur:",
    "CONTACT_EMAIL": "Courrier électronique:",
    "CONTACT_GITHUB": "Github:",
    "CONTACT_LINKEDIN": "LinkedIn:",
    "FLAG_ENGLISH_HINT": "Anglais",
    "FLAG_FRENCH_HINT": "Français",
    "FLAG_SPANISH_HINT": "Espagnol",
    "GAME": "Jeu",
    "GAME_OVER": "La partie est finie",
    "GAME_STARTED_BY": "Commencé par",
    "GAME_STARTED_BY_COMPUTER": "Ordinateur",
    "GAME_STARTED_BY_YOU": "Toi",
    "GAME_DELETING": "En train de supprimer le jeu de la bsae de données...",
    "GAME_DELETE_ERROR": "Une erreur s'est produite lors de la suppression du jeu de la base de données.",
    "GAME_SAVING": "En train de sauvegarder les données du jeu...",
    "GAME_SAVE_ERROR": "Une erreur s'est produite lors de l'enregistrement des résultats du jeu dans la base de données.",
    "HISTORICAL": "Historique",
    "HISTORICAL_NO_DATA_YET": "Il n'y a pas encore d'historique. Reviens un peu plus tard après avoir joué quelques parties.",
    "HISTORICAL_FETCHING_GAMES": "Récupération des jeux...",
    "HISTORICAL_FETCHING_GAME": "Récupération du jeu...",
    "HISTORICAL_FETCHING_GAMES_ERROR": "Une erreur s'est produite lors de la récupération des jeux de la base de données.",
    "HISTORICAL_FETCHING_GAME_ERROR": "Une erreur s'est produite lors de la récupération du jeu de la base de données.",
    "IMAGE_COMPUTER_LABEL": "Ordi",
    "IMAGE_COMPUTER_TOOLTIP": "L'ordinateur va jouer avec cette image.",
    "IMAGE_REFRESH_TOOLTIP": "Fais clic à fin de jouer avec une autre image.",
    "IMAGE_USER_LABEL": "Toi",
    "IMAGE_USER_TOOLTIP": "Tu vas jouer avec cette image.",
    "LOGIN": "Se connecter",
    "MESSAGE_DRAW": "Match nul!",
    "MESSAGE_LOST": "Tu as perdu!",
    "MESSAGE_WON": "Tu as gagné!",
    "PLAYING_LEVEL_EASY": "Facile",
    "PLAYING_LEVEL_LABEL": "Niveau de jeu",
    "PLAYING_LEVEL_SMART": "Difficile",
    "PLAYING_LEVEL_TOOLTIP": "Quelle est la réflexion de l'ordinateur avant de jouer.",
    "REPOSITORY": "Banque",
    "SIGNUP": "S'inscrire",
    "WELCOME_BACK": "Ravis de vous retrouver",
    "START_GAME_LABEL": "Premier",
    "START_GAME_ANSWER_NO": "Non",
    "START_GAME_TOOLTIP": "Qui commence la partie: L'ordinateur ou toi?",
    "START_GAME_ANSWER_YES": "Oui"
  }
]
