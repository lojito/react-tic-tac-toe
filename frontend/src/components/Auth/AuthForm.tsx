import React, { FC, useContext, useEffect, useRef, useState } from "react";
import { DictionaryContext } from "../../contexts/dictionary/DictionaryContext";
import "./Auth.scss";

export type Account = {
  email: string;
  name?: string;
  password: string;
};

interface Props {
  onSignup?: (account: Account) => void;
  onLogin?: (account: Account) => void;
  authType: string;
  error: string;
}

const AuthForm: FC<Props> = ({ onSignup, onLogin, authType, error }) => {
  const {
    dictionary: {
      BUTTON_SIGNUP,
      BUTTON_SIGNIN,
      AUTH_FORM_EMAIL_LABEL,
      AUTH_FORM_EMAIL_PLACEHOLDER,
      AUTH_FORM_NAME_LABEL,
      AUTH_FORM_NAME_PLACEHOLDER,
      AUTH_FORM_PASSWORD_PLACEHOLDER,
      AUTH_FORM_PASSWORD_LABEL,
    },
  } = useContext(DictionaryContext);

  const [account, setAccount] = useState<Account>({
    email: "",
    name: "",
    password: "",
  });

  const emailRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    emailRef!.current!.focus();
  }, []);

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (authType === "signup") {
      onSignup!(account);
    } else {
      onLogin!(account);
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setAccount({
      ...account,
      [e.target.id]: e.target.value,
    });
  };

  const btnName = authType === "signup" ? BUTTON_SIGNUP : BUTTON_SIGNIN;

  return (
    <section className="authSection">
      <form onSubmit={handleSubmit}>
        <div className="input">
          <label htmlFor="email">{AUTH_FORM_EMAIL_LABEL}</label>
          <input
            id="email"
            type="email"
            autoComplete="email"
            required
            value={account.email}
            placeholder={AUTH_FORM_EMAIL_PLACEHOLDER}
            onChange={handleChange}
            ref={emailRef}
          />
        </div>

        {authType === "signup" && (
          <div className="input">
            <label htmlFor="name">{AUTH_FORM_NAME_LABEL}</label>
            <input
              id="name"
              type="text"
              required
              value={account.name}
              placeholder={AUTH_FORM_NAME_PLACEHOLDER}
              onChange={handleChange}
            />
          </div>
        )}

        <div className="input">
          <label htmlFor="password">{AUTH_FORM_PASSWORD_LABEL}</label>
          <input
            id="password"
            type="password"
            autoComplete="current-password"
            required
            value={account.password}
            placeholder={AUTH_FORM_PASSWORD_PLACEHOLDER}
            onChange={handleChange}
          />
        </div>

        <div className="button">
          <button type="submit" name={btnName}>
            {btnName}
          </button>
        </div>

        {error && (
          <div className="info" data-testid="error">
            {error}
          </div>
        )}
      </form>
    </section>
  );
};

export default AuthForm;
