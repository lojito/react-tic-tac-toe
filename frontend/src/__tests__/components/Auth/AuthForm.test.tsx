import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import AuthForm from "../../../components/Auth/AuthForm";
import { DictionaryContext } from "../../../contexts/dictionary/DictionaryContext";

jest.mock("../../../hooks/auth/useAuth");

describe("AuthForm", () => {
  const handleSignup = jest.fn();
  const handleLogin = jest.fn();

  const history = createMemoryHistory();
  history.push = jest.fn();

  const mockDictionary = {
    AUTH_FORM_EMAIL_LABEL: "E-mail",
    AUTH_FORM_NAME_LABEL: "Name",
    AUTH_FORM_PASSWORD_LABEL: "Password",
    BUTTON_SIGNUP: "Sign up",
    BUTTON_SIGNIN: "Log in",
  };

  interface Props {
    authType?: string;
    error?: string;
  }

  const TestComponent = ({ authType = "signup", error = "" }: Props) => {
    return (
      <DictionaryContext.Provider value={{ dictionary: mockDictionary }}>
        <Router history={history}>
          <AuthForm
            onSignup={handleSignup}
            onLogin={handleLogin}
            authType={authType}
            error={error}
          />
        </Router>
      </DictionaryContext.Provider>
    );
  };

  const renderAuthForm = ({ authType = "signup", error = "" }: Props = {}) => {
    const utils = render(<TestComponent authType={authType} error={error} />);

    const emailLabel = screen.getByLabelText(
      mockDictionary.AUTH_FORM_EMAIL_LABEL
    );
    const email = screen.getByRole("textbox", {
      name: mockDictionary.AUTH_FORM_EMAIL_LABEL,
    });

    const nameLabel = screen.queryByLabelText(
      mockDictionary.AUTH_FORM_NAME_LABEL
    );
    const name = screen.queryByRole("textbox", {
      name: mockDictionary.AUTH_FORM_NAME_LABEL,
    });

    const passwordLabel = screen.getByLabelText(
      mockDictionary.AUTH_FORM_PASSWORD_LABEL
    );

    const signUpButton = screen.queryByRole("button", {
      name: mockDictionary.BUTTON_SIGNUP,
    });
    const signInButton = screen.queryByRole("button", {
      name: mockDictionary.BUTTON_SIGNIN,
    });

    return {
      ...utils,
      mockDictionary,
      handleSignup,
      handleLogin,
      emailLabel,
      email,
      nameLabel,
      name,
      passwordLabel,
      signUpButton,
      signInButton,
    };
  };

  it("should render correctly", () => {
    const { asFragment } = renderAuthForm();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the signup form", () => {
    const {
      emailLabel,
      email,
      nameLabel,
      name,
      passwordLabel,
      signUpButton,
      signInButton,
    } = renderAuthForm();

    expect(emailLabel).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(nameLabel).toBeInTheDocument();
    expect(name).toBeInTheDocument();
    expect(passwordLabel).toBeInTheDocument();
    expect(signUpButton).toBeInTheDocument();
    expect(signInButton).not.toBeInTheDocument();
  });

  it("should render the login form", () => {
    const {
      emailLabel,
      email,
      nameLabel,
      name,
      passwordLabel,
      signUpButton,
      signInButton,
    } = renderAuthForm({ authType: "login" });

    expect(emailLabel).toBeInTheDocument();
    expect(email).toBeInTheDocument();
    expect(nameLabel).not.toBeInTheDocument();
    expect(name).not.toBeInTheDocument();
    expect(passwordLabel).toBeInTheDocument();
    expect(signUpButton).not.toBeInTheDocument();
    expect(signInButton).toBeInTheDocument();
  });

  it("should call the auth handler when signign up the user", async () => {
    const user = userEvent.setup();
    const { signUpButton, handleSignup } = renderAuthForm();

    await user.click(signUpButton!);

    expect(handleSignup).toHaveBeenCalled();
  });

  it("should call the auth handler when loging in the user", async () => {
    const user = userEvent.setup();
    const { signInButton, handleLogin } = renderAuthForm({ authType: "login" });

    await user.click(signInButton!);

    expect(handleLogin).toHaveBeenCalled();
  });

  it("should error when signing up the user", async () => {
    const user = userEvent.setup();
    const { signUpButton, rerender } = renderAuthForm();

    await user.click(signUpButton!);

    rerender(
      <TestComponent error="There was an error while trying to sign up the user" />
    );

    const errorDiv = screen.queryByTestId("error");

    expect(errorDiv).toBeInTheDocument();
  });

  it("should error when loging in the user", async () => {
    const user = userEvent.setup();
    const { signInButton, rerender } = renderAuthForm({ authType: "login" });

    await user.click(signInButton!);

    rerender(
      <TestComponent
        authType="login"
        error="There was an error while trying to log in the user"
      />
    );

    const errorDiv = screen.queryByTestId("error");

    expect(errorDiv).toBeInTheDocument();
  });

  it("should not error when loging in the user", async () => {
    const user = userEvent.setup();
    const { signInButton, rerender } = renderAuthForm({ authType: "login" });

    await user.click(signInButton!);

    rerender(<TestComponent authType="login" />);
    const errorDiv = screen.queryByTestId("error");

    expect(errorDiv).not.toBeInTheDocument();
  });

  it("should allow to enter the email", async () => {
    const user = userEvent.setup();
    const userEmail = "livanojito@gmail.com";
    const { email } = renderAuthForm({ authType: "login" });

    await user.type(email, userEmail);

    await waitFor(() => {
      expect(email).toHaveValue(userEmail);
    });
  });

  it("should allow to enter the name", async () => {
    const user = userEvent.setup();
    const userName = "livan";
    const { name } = renderAuthForm();

    await user.type(name!, userName);

    await waitFor(() => {
      expect(name).toHaveValue(userName);
    });
  });

  it("should allow to enter the password", async () => {
    const user = userEvent.setup();
    const userPassword = "password";
    const { passwordLabel } = renderAuthForm();

    await user.type(passwordLabel!, userPassword);

    await waitFor(() => {
      expect(passwordLabel).toHaveValue(userPassword);
    });
  });
});
