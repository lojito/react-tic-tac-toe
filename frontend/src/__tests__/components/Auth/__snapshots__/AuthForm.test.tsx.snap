// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm should render correctly 1`] = `
<DocumentFragment>
  <section
    class="authSection"
  >
    <form>
      <div
        class="input"
      >
        <label
          for="email"
        >
          E-mail
        </label>
        <input
          autocomplete="email"
          id="email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="input"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          id="name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div
        class="input"
      >
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          id="password"
          required=""
          type="password"
          value=""
        />
      </div>
      <div
        class="button"
      >
        <button
          name="Sign up"
          type="submit"
        >
          Sign up
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
